[27215] PullPreview: Share "files" directory from "web" with "worker"

wielinde opened the pull request:
https://community.openproject.com/projects/packaging-and-distribution/work_packages/27215/activity

wielinde committed:
This is a try to expose ‚files‘ directory from web and hook it into worker

fix order and mixins

do not use a default block that starts its own servers.

crohr commented on docker-compose.pullpreview.yml:
```diff
@@ -23,12 +21,25 @@ worker: &ruby
   command: "./docker/web"
   memory: 1024
+  volumes:
+    - .:/files
```
@wielinde Just keep `.:/usr/src/app` here instead of `.:/files`, and then both `worker` and `web` processes will share the same `/usr/src/app` directory (so if you have `/usr/src/app/files` in the `web` container instance, it will be exactly the same shared volume in the `worker` container instance) !

wielinde committed:
read full paths for volumes. `files` as sibling of `app`

wielinde committed:
Seems like /usr/src/app is the folder where everything is about…

crohr committed:
share only /usr/src/app/files directory, not persisted

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b8a6c857bfbad687c50a9f27d274e04c5d1e4297) (deployed commit: d9894a1447c0688474ff6056b89c85fd3c9f3baf).

crohr committed:
Fix Dockerfile

crohr committed:
fix

crohr committed:
More involved than expected, since we were not starting the container as root.

So, adding a default entrypoint that switches to the APP_USER, and chown the APP_DATA folder to make sure it has the correct permissions.
(By default, docker creates volumes owned by root, so in the previous version the APP_USER couldn't write to the folder).
This is what's done in postgres: https://github.com/docker-library/postgres/blob/2f0f505894f10c2e65a38591ec0cd417dbcc03e6/10/docker-entrypoint.sh

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/4fc2292b8fadbc5d89f71cfbb58b02f109e8ab00) (deployed commit: cc95391a69c52437fd244a4191d42eaf7c12ee42).

crohr commented:
@wielinde it was more involved than what I expected, because shared volumes in docker are by default created with `root` permissions, so the processes couldn't write to it since we explicitly started them with `USER app`.

I fixed this by removing all references to the `app` user in the Dockerfile, and just switching to that user when starting the container (as is also done by e.g. the postgres docker image). This allows to properly set the permissions on various folders in the container instance. A nice side-effect being that it simplifies the Dockerfile, and doesn't require as many layers (and thus disk usage) as before.

The only issue is that this would need to be tested against the downstream projects that reuse the Dockerfile to make sure it doesn't break them, as well as testing whether the Dockerfile.public is still working as expected (the latter I can do). /cc @machisuji 

In the meantime I'll let you test the running PullPreview instance, and let me know if you find anything weird.

oliverguenther commented:
Only FYI I'm killing the travis builds when necessary, the docker files aren't used there anyway so testing that is not necessary.

crohr committed:
Simplify [ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/16652273a6520e39bf572203b0affec75fbe1339) (deployed commit: 02116c07116ab2ee0c009b88e55d17e2ffee3c84).
